File: snakeandladder.rb


MAXBOARDPOSITION = 30
snake_and_ladder = Hash.new

snake_and_ladder[27] = 1
snake_and_ladder[11] = 26
snake_and_ladder[3] = 22
snake_and_ladder[21] = 9
snake_and_ladder[17] = 4
snake_and_ladder[5] = 8
snake_and_ladder[19] = 7
snake_and_ladder[20] = 29


Player = Struct.new(:name, :position)

a = Player.new("Vimlesh",0)
b = Player.new("Computer",0)
c = Player.new("Rahul",0)

players =[a,b,c]

def throw_dice
	r = Random.new()
	r.rand(1..6)
end


def play_game(players, snake_and_ladder)	
	catch(:done) do
		while true
			players.each{
				|player|
				dice_value = throw_dice
				next_position = player.position + dice_value


				case 
				when  next_position > MAXBOARDPOSITION
					puts(" #{player.name} can't proceed... waiting for #{MAXBOARDPOSITION - player.position}")
				when next_position < MAXBOARDPOSITION
					snl_position = snake_and_ladder[next_position]
					
					player.position = next_position
					if snl_position
						puts(" Gotcha for player #{player.name} @ #{next_position} Moving from #{next_position} -> #{snl_position}")
						next_position = snl_position
						
						player.position = next_position
					# else
					# 	player.position = next_position
					end
					puts(" #{player.name} is at new position #{player.position} and rolled #{dice_value}")

				when next_position == MAXBOARDPOSITION
					player.position = next_position
					puts("#{player.name} is at new position #{player.position} and rolled is #{dice_value} and wins <----")
	                throw :done
				end
			}
			sleep(0.1)
		end
	end
end


play_game(players, snake_and_ladder)

File: snakeandladder.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

var snakesAndLader map[int]int

const MaxBoardPosition = 30

type Player struct {
	Name     string
	Position int
}

func main() {
	snakesAndLader = make(map[int]int)
	snakesAndLader[27] = 1
	snakesAndLader[11] = 26
	snakesAndLader[3] = 22
	snakesAndLader[21] = 9
	snakesAndLader[17] = 4
	snakesAndLader[5] = 8
	snakesAndLader[19] = 7
	snakesAndLader[20] = 29

	A := Player{Name: "Vimlesh"}
	B := Player{Name: "Sunil"}
	C := Player{Name: "Rahul"}
	D := Player{Name: "Computer"}
		

	PlayGame(A, B, C, D)
}

func ThrowDice() int {
	rand.Seed(time.Now().UnixNano())
	diceVal := rand.Intn(7)
	if diceVal == 0 {
		diceVal++
	}
	return diceVal
}

func PlayGame(players ...Player) {

loop:
	for {
		for index, player := range players {

			diceValue := ThrowDice()
			nextPosition := player.Position + diceValue

			switch {
			case nextPosition > MaxBoardPosition:
				fmt.Printf(" %s can't proceed... waiting for %d \n", player.Name, MaxBoardPosition-player.Position)
			case nextPosition < MaxBoardPosition:
				snl_pos, found := snakesAndLader[nextPosition]
				if found {
					fmt.Printf(" Gotcha for player %s @ %d Moving from %d -> %d \n", player.Name, nextPosition, nextPosition, snl_pos)
					nextPosition = snl_pos
					player.Position = nextPosition
				} else {
					player.Position = nextPosition
				}

				players[index] = player
				fmt.Printf(" %s is at new position %d and rolled %d  \n", player.Name, player.Position, diceValue)

			case nextPosition == MaxBoardPosition:
				player.Position = nextPosition
				players[index] = player
				fmt.Printf(" %s is at new position %d and rolled is %d and wins <---- \n", player.Name, player.Position, diceValue)

				break loop
			}

			time.Sleep(10 * time.Millisecond)
		}
		fmt.Println("")
	}
}
